### PR TITLE
jit: refactor around builder.

### PR DESCRIPTION
**Pull request by mathetake**
This commit Introduces `compiler` interface as the architecture specific 
native code compiler (previously called amd64Builder as an implementation for it).
As a result, this also moves the `compilerWasmFunction` in engine.go by making it
architecture-independent via `compiler` interface.

### CODE REVIEW

File: wasm/jit/engine.go [hunk 1]
```diff
@@ -446,3 +447,230 @@ func (e *engine) builtinFunctionMemoryGrow(f *compiledWasmFunction) {
 func (e *engine) builtinFunctionMemorySize(f *compiledWasmFunction) {
 	e.push(uint64(len(f.memory.Buffer)) / wasm.PageSize)
 }
+
+func (e *engine) compileWasmFunction(f *wasm.FunctionInstance) (*compiledWasmFunction, error) {
```
Comment (mathetake) [inline wasm/jit/engine.go:451]: note: moved from jit_amd64.go as the architecture-independent form via `compiler` interface

File: wasm/jit/engine.go [hunk 2]
```diff
@@ -446,3 +447,230 @@ func (e *engine) builtinFunctionMemoryGrow(f *compiledWasmFunction) {
...
+	}
+
+	compiler.emitPreamble()
+
+	for _, op := range ir.Operations {
+		switch o := op.(type) {
+		case *wazeroir.OperationUnreachable:
+			compiler.compileUnreachable()
+		case *wazeroir.OperationLabel:
+			if err := compiler.compileLabel(o); err != nil {
```
Comment (codefromthecrypt) [inline wasm/jit/engine.go:469]: what if you made err a variable then if set wrapped it based on the string of op?

File: wasm/jit/engine.go [hunk 3]
```diff
@@ -446,3 +447,230 @@ func (e *engine) builtinFunctionMemoryGrow(f *compiledWasmFunction) {
...
+			if err := compiler.compileLoad8(o); err != nil {
+				return nil, fmt.Errorf("error handling load8 operation: %w", err)
+			}
+		case *wazeroir.OperationLoad16:
+			if err := compiler.compileLoad16(o); err != nil {
+				return nil, fmt.Errorf("error handling load16 operation: %w", err)
+			}
+		case *wazeroir.OperationLoad32:
+			if err := compiler.compileLoad32(o); err != nil {
+				return nil, fmt.Errorf("error handling load16 operation: %w", err)
```
Comment (codefromthecrypt) [inline wasm/jit/engine.go]: I think the formatting can be templated

```suggestion
				return nil, fmt.Errorf("error handling load32 operation: %w", err)
```

File: wasm/jit/engine.go [hunk 4]
```diff
@@ -446,3 +447,230 @@ func (e *engine) builtinFunctionMemoryGrow(f *compiledWasmFunction) {
...
+		case *wazeroir.OperationExtend:
+			return nil, fmt.Errorf("unsupported operation in JIT compiler: %v", o)
+		default:
+			return nil, fmt.Errorf("unreachable: a bug in JIT compiler")
+		}
+	}
+
+	code, maxStackPointer, err := compiler.compile()
+	if err != nil {
+		return nil, fmt.Errorf("failed to assemble: %w", err)
```
Comment (codefromthecrypt) [inline wasm/jit/engine.go:609]:
```suggestion
		return nil, fmt.Errorf("failed to compile: %w", err)
```

File: wasm/jit/engine.go [hunk 5]
```diff
@@ -446,3 +447,230 @@ func (e *engine) builtinFunctionMemoryGrow(f *compiledWasmFunction) {
...
+	}
+
+	cf := &compiledWasmFunction{
+		source:          f,
+		codeSegment:     code,
+		params:          uint64(len(f.Signature.ParamTypes)),
```
Comment (codefromthecrypt) [inline wasm/jit/engine.go:615]:
```suggestion
		paramCount:          uint64(len(f.Signature.ParamTypes)),
```

File: wasm/jit/engine.go [hunk 6]
```diff
@@ -446,3 +447,230 @@ func (e *engine) builtinFunctionMemoryGrow(f *compiledWasmFunction) {
...
+		results:         uint64(len(f.Signature.ResultTypes)),
```
Comment (codefromthecrypt) [inline wasm/jit/engine.go:616]:
```suggestion
		resultCount:         uint64(len(f.Signature.ResultTypes)),
```

File: wasm/jit/engine.go [hunk 7]
```diff
@@ -446,3 +447,182 @@ func (e *engine) builtinFunctionMemoryGrow(f *compiledWasmFunction) {
...
+		case *wazeroir.OperationI32ReinterpretFromF32,
+			*wazeroir.OperationI64ReinterpretFromF64,
+			*wazeroir.OperationF32ReinterpretFromI32,
+			*wazeroir.OperationF64ReinterpretFromI64:
+			err = fmt.Errorf("unsupported operation")
+		case *wazeroir.OperationExtend:
+			err = fmt.Errorf("unsupported operation")
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to compile operation %s: %w", op.Kind().String(), err)
```
Comment (mathetake) [inline wasm/jit/engine.go:603]: e.g.: `failed to compile Wasm function: failed to compile operation Ge: unsupported operation`

File: wasm/jit/jit_amd64.go [hunk 8]
```diff
@@ -297,16 +57,26 @@ type amd64Builder struct {
 	requireFunctionCallReturnAddressOffsetResolution []*obj.Prog
 }
 
-func (b *amd64Builder) compile() ([]byte, error) {
-	code, err := mmapCodeSegment(b.builder.Assemble())
+// Implements compiler.emitPreamble for amd64Compiler.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:60]: I think godoc like this don't add value. Also IDEs warn when they don't start with the function name

File: wasm/jit/jit_amd64.go [hunk 9]
```diff
@@ -297,16 +57,26 @@ type amd64Builder struct {
...
+func (c *amd64Compiler) emitPreamble() {
+	// We assume all function parameters are already pushed onto the stack by
+	// the caller.
+	c.pushFunctionParams()
+	// Initialize the reserved registers first of all.
+	c.initializeReservedRegisters()
+}
+
+// Implements compiler.compile for amd64Compiler.
+func (c *amd64Compiler) compile() ([]byte, uint64, error) {
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:70]: I think godoc like this don't add value. Also IDEs warn when they don't start with the function name

File: wasm/jit/engine.go (hunk 2 above)
Comment (mathetake) [reply wasm/jit/engine.go]: consolidated the error handling! thanks!

**Review by codefromthecrypt — APPROVED** [review]
ignore any already addressed stuff. I like the naming so much better!

**Comment by mathetake** [comment]
Thanks for the reviews!

**Pull request merged**